ligo/src/registryDAO.mligo

```diff
@@ -0,0 +1,190 @@
+// SPDX-FileCopyrightText: 2020 TQ Tezos
+// SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+#include "types.mligo"
+#include "proposal.mligo"
+#include "registryDAO/types.mligo"
+
+(*
+ * This is a dummy entrypoint so that this file can be called with LIGO's `compile-storage` command as shown below, providing admin and token addresses.
+ *
+ * ligo compile-storage ligo/src/registryDAO.mligo dummy 'default_registry_DAO_full_storage(("tz1QozfhaUW4wLnohDo6yiBUmh7cPCSXE9Af" : address), ("tz1QozfhaUW4wLnohDo6yiBUmh7cPCSXE9Af" : address), 0n, 0n, 0n, 0n, 0n)'
+ *)
+let dummy (action, store : parameter * full_storage) : operation list * full_storage
+  = (([] : operation list), store)
+
+let lookup_config (key_name, ce : string * contract_extra) : nat option =
+    match Map.find_opt key_name ce with
+      Some (packed_b) ->
+        begin
+          match ((Bytes.unpack packed_b) : ((nat option) option)) with
+            Some (v) -> v
+          | None -> (failwith "CONFIG_UNPACKING_FAILED" : nat option)
+        end
+    | None -> (failwith "CONFIG_VALUE_NOT_FOUND" : nat option)
+
+let require_config (key_name, ce : string * contract_extra) : nat =
+  match lookup_config (key_name, ce) with
+    Some (v) -> v
+  | None -> (failwith "REQUIRED_CONFIG_NOT_FOUND" : nat)
+
+let apply_diff_registry (diff, registry
+  : ((registry_key, registry_value option) map) * registry) : registry =
+  let
+    foldFn (registry, update: registry * (registry_key * (registry_value option))) : registry
+      = Map.update update.0 update.1 registry
+  in Map.fold foldFn diff registry
+
+let apply_diff_registry_affected (proposal_key, diff, registry_affected : proposal_key
+        * ((registry_key, registry_value option) map) * registry_affected) : registry_affected =
+  let
+    foldFn (registry_affected, update: registry_affected * (registry_key *
+      (registry_value option))) : registry_affected =
+  Map.update update.0 (Some (proposal_key)) registry_affected in
+    Map.fold foldFn diff registry_affected
+
+let apply_diff (diff, ce : ((registry_key, registry_value option) map) * contract_extra)
+  : contract_extra =
+  match Map.find_opt "registry" ce with
+    Some (packed_registry) ->
+      begin
+        match (Bytes.unpack (packed_registry) : registry option) with
+          Some (registry) -> let
+            new_registry : registry = apply_diff_registry (diff, registry)
+            in Map.update "registry" (Some (Bytes.pack new_registry)) ce
+        | None -> (failwith "REGISTRY_NOT_FOUND" : contract_extra)
+      end
+  | None -> (failwith "REGISTRY_NOT_FOUND" : contract_extra)
```

Comment:
Why not replace this outer error with `require_config` as well?
I think the generic error is fine.

Sadly I cannot think of a way to de-duplicate the logic of `Bytes.unpack` on the spot because of the type signature requirement :thinking:

Reply:
Originally, the `require_config` only worked with `nat option`, but here we needed a map. But I have now refactored it with some helpers that this verbosity is a bit reduced.